test/cpp/util/channelz_sampler_test.cc

```diff
@@ -0,0 +1,181 @@
+
+  // forces channelz and channel tracing to be enabled.
+  builder.AddChannelArgument(GRPC_ARG_ENABLE_CHANNELZ, 1);
+  builder.AddChannelArgument(GRPC_ARG_MAX_CHANNEL_TRACE_EVENT_MEMORY_PER_NODE,
+                             1024);
+  builder.RegisterService(&service);
+  std::unique_ptr<Server> server(builder.BuildAndStart());
+  gpr_log(GPR_INFO, "Server listening on %s", server_address.c_str());
+  while (true) {
+    if (gpr_event_get(done_ev)) {
+      return;
+    }
+  }
+}
+
+// Run client in a thread - - set timeout as 0.15s
+void RunClient(std::string client_id, gpr_event* done_ev) {
+  grpc::ChannelArguments channel_args;
+  std::shared_ptr<grpc::ChannelCredentials> channel_creds =
+      grpc::testing::GetCredentialsProvider()->GetChannelCredentials(
+          custom_credentials_type, &channel_args);
+  std::unique_ptr<grpc::testing::TestService::Stub> stub =
+      grpc::testing::TestService::NewStub(
+          grpc::CreateChannel(server_address, channel_creds));
+  unsigned int client_echo_sleep_second = 1;
+
+  gpr_log(GPR_INFO, "Client %s is echoing!", client_id.c_str());
+  while (true) {
+    if (gpr_event_get(done_ev)) {
+      return;
+    }
+    sleep(client_echo_sleep_second);
+    // Rcho RPC
+    grpc::testing::Empty request;
+    grpc::testing::Empty response;
+    ClientContext context;
+    int64_t timeout_microseconds = 150;
+    context.set_deadline(
+        grpc_timeout_milliseconds_to_deadline(timeout_microseconds));
+    stub->EmptyCall(&context, request, &response);
+  }
+}
+
+// Create the channelz to test the connection to the server
+void WaitForConnection() {
+  grpc::ChannelArguments channel_args;
+  std::shared_ptr<grpc::ChannelCredentials> channel_creds =
+      grpc::testing::GetCredentialsProvider()->GetChannelCredentials(
+          custom_credentials_type, &channel_args);
+  auto channel = grpc::CreateChannel(server_address, channel_creds);
+  channel->WaitForConnected(grpc_timeout_seconds_to_deadline(3));
+}
+
+// Test the channelz sampler
+TEST(ChannelzSamplerTest, SimpleTest) {
+  // server thread
+  gpr_event done_ev0;
+  gpr_event_init(&done_ev0);
+  std::thread server_thread(RunServer, &done_ev0);
+  WaitForConnection();
+
+  // client threads
+  gpr_event done_ev1, done_ev2;
+  gpr_event_init(&done_ev1);
+  gpr_event_init(&done_ev2);
+  std::thread client_thread_1(RunClient, "1", &done_ev1);
+  std::thread client_thread_2(RunClient, "2", &done_ev2);
+
+  // Run the channelz sampler
+  std::string channelz_sampler_bin_path =
+      "./bazel-bin/test/cpp/util/channelz_sampler";
+  grpc::SubProcess* test_driver = new grpc::SubProcess(
+      {std::move(channelz_sampler_bin_path),
+       "--server_address=" + server_address,
+       "--custom_credentials_type=" + custom_credentials_type,
+       "--sampling_times=" + sampling_times,
+       "--sampling_interval_seconds=" + sampling_interval_seconds,
+       "--output_json=" + output_json});
+
+  test_driver->Join();
```

Comment:
there are still no assertions being made to check that the subprocess didn't fail -- please add

Reply:
done